src/System.Linq.Parallel/tests/QueryOperators/ForAllTests.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using Xunit;
+
+namespace Test
+{
+    public class ForAllTests
+    {
+        [Theory]
+        [MemberData("Ranges", (object)(new int[] { 1, 2, 16 }), MemberType = typeof(Sources))]
+        [MemberData("Ranges", (object)(new int[] { 1, 2, 16 }), MemberType = typeof(UnorderedSources))]
+        public static void ForAll(Labelled<ParallelQuery<int>> labelled, int count)
+        {
+            ParallelQuery<int> query = labelled.Item;
+            IntegerRangeSet seen = new IntegerRangeSet(0, count);
+            query.ForAll<int>(x => Assert.True(seen.Add(x)));
```

Comment:
Okay, looks like locking (ReaderWriterLockSlim) is about the same, +1%.
Moral of the story: IEnumerable.ElementAt(...) isn't guaranteed to work in O(n) time over BitArrays (probably because LINQ doesn't have special handling for it).  Of course, it doesn't help it's not exposing AsQueryable at the moment either...